homeassistant/components/tahoma.py

```diff
@@ -0,0 +1,1166 @@
+"""
+Support for Tahoma devices.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tahoma/
+"""
+import logging
+import json
+import requests
+import voluptuous as vol
+
+from collections import defaultdict
+from homeassistant.util import (convert, slugify)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.helpers.entity import Entity
+from homeassistant.exceptions import HomeAssistantError
+
+TAHOMA_CONTROLLER = None
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'tahoma'
+_BASE_URL = 'https://www.tahomalink.com/enduser-mobile-web/externalAPI/json/'
+_BASE_HEADERS = {'User-Agent': 'mine'}
+CONF_EXCLUDE = 'exclude'
+
+TAHOMA_ID_LIST_SCHEMA = vol.Schema([cv.string])
+
+TAHOMA_DEVICES = defaultdict(list)
+TAHOMA_ID_FORMAT = '{}_{}'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_EXCLUDE, default=[]): TAHOMA_ID_LIST_SCHEMA,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+def setup(hass, config):
+    """docstring for async_setup"""
+    global TAHOMA_CONTROLLER
+    conf = config[DOMAIN]
+    username = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
+    exclude = conf.get(CONF_EXCLUDE)
+
+    TAHOMA_CONTROLLER = TahomaApi(username, password)
+    TAHOMA_DEVICES['api'] = TAHOMA_CONTROLLER
+
+    TAHOMA_CONTROLLER.getSetup()
+    devices = TAHOMA_CONTROLLER.getDevices()
+    #user = TAHOMA_CONTROLLER.getUser()
+    #actionGroups = TAHOMA_CONTROLLER.getActionGroups()
+
+    for device in devices:
+        d = TAHOMA_CONTROLLER.getDevice(device)
+        _LOGGER.error(d.label)
+        _LOGGER.error(d.type)
+        if any(ext not in d.type for ext in exclude):
+            device_type = map_tahoma_device(d)
+            if device_type is None:
+                continue
+            TAHOMA_DEVICES[device_type].append(d)
+
+    return True
+
+def map_tahoma_device(tahoma_device):
+    """Map tahoma classes to Home Assistant types."""
+    if tahoma_device.type.lower().find("shutter") != -1:
+        return 'cover'
+    elif tahoma_device.type == 'io:LightIOSystemSensor':
+        return 'sensor'
+    return None
+
+class TahomaApi:
```

Comment:
This should be extracted to a stand-alone Python module which can be imported. This way other projects can use the code to interact with Tahoma devices.